src/main/scala/akka/persistence/jdbc/query/JournalSequenceActor.scala

```diff
@@ -57,47 +57,34 @@ class JournalSequenceActor(readJournalDao: ReadJournalDao, config: JournalSequen
    */
   def receive(currentMaxOrdering: OrderingId, missingByCounter: Map[Int, Set[OrderingId]], moduloCounter: Int, previousDelay: FiniteDuration = queryDelay): Receive = {
+
     case ScheduleAssumeMaxOrderingId(max) =>
       // All elements smaller than max can be assumed missing after this delay
       val delay = queryDelay * maxTries
       timers.startSingleTimer(key = AssumeMaxOrderingIdTimerKey, AssumeMaxOrderingId(max), delay)
+
     case AssumeMaxOrderingId(max) =>
       if (currentMaxOrdering < max) {
         context.become(receive(max, missingByCounter, moduloCounter, previousDelay))
       }
+
     case GetMaxOrderingId =>
       sender() ! MaxOrderingId(currentMaxOrdering)
+
     case QueryOrderingIds =>
-      readJournalDao.journalSequence(currentMaxOrdering, batchSize).runWith(Sink.seq)
-        .map(result => NewOrderingIds(result)) pipeTo self
-    case NewOrderingIds(elements) =>
-      val givenUp = missingByCounter.getOrElse(moduloCounter, Set.empty)
-      val (nextmax, _, missingElems) = elements.foldLeft[(OrderingId, OrderingId, Set[OrderingId])](currentMaxOrdering, currentMaxOrdering, Set.empty) {
-        case ((currentMax, previousElement, missing), elem) =>
-          val newMissing = if (previousElement + 1 < elem && !givenUp(elem)) {
-            val currentlyMissing = previousElement + 1 until elem
-            def alreadyMissing(e: Long) = missingByCounter.values.exists(_.contains(e))
-            missing ++ currentlyMissing.filterNot(alreadyMissing)
-          } else missing
-          val newMax =
-            if (currentMax + 1 == elem) elem
-            else if ((currentMax + 1).until(elem).forall(givenUp.contains)) elem
-            else currentMax
-          (newMax, elem, newMissing)
-      }
-      val newMissingByCounter = (missingByCounter + (moduloCounter -> missingElems)).map {
-        case (key, value) =>
-          key -> value.filter(missingId => missingId > nextmax)
-      }
-      if (nextmax - currentMaxOrdering >= batchSize && newMissingByCounter.values.forall(_.isEmpty)) {
-        // Many elements have been retrieved but none are missing
-        // We can query again immediately, as this allows the actor to rapidly retrieve the real max ordering
-        self ! QueryOrderingIds
-        context.become(receive(nextmax, newMissingByCounter, moduloCounter))
-      } else {
-        scheduleQuery(queryDelay)
-        context.become(receive(nextmax, newMissingByCounter, (moduloCounter + 1) % maxTries))
-      }
+      readJournalDao
+        .journalSequence(currentMaxOrdering, batchSize)
+        .runWith(Sink.seq)
+        .map(result => NewOrderingIds(currentMaxOrdering, result)) pipeTo self
+
+    case NewOrderingIds(originalOffset, _) if originalOffset < currentMaxOrdering =>
+      // search was done using an offset that became obsolete in the meantime
+      // therefore we start a new query
+      self ! QueryOrderingIds
```

Comment:
This cover the case when the `currentMaxOrdering` has been updated by `ScheduleAssumeMaxOrderingId`